simperby/src/node/stf.rs

```diff
@@ -0,0 +1,288 @@
+                            last_header.version
+                        ))
+                    })?;
+                    let ver2 = semver::Version::parse(version).map_err(|_| {
+                        ExecutionError::InvalidHeader(format!(
+                            "invalid version: {}",
+                            last_header.version
+                        ))
+                    })?;
+                    if ver2 <= ver1 {
+                        return Err(ExecutionError::InvalidHeader(format!(
+                            "version must grow: {}",
+                            ver2
+                        )));
+                    }
+                    next_version = ver2.to_string();
+                }
+            }
+        }
+    }
+    let merkle_tree = simperby_merkle_tree::oneshot::OneshotMerkleTree::create(
+        transactions.iter().map(|t| t.hash()).collect(),
+    );
+    let delegation_state = storage.get_delegation_state().await?;
+    Ok(BlockHeader {
+        author: header_info.author,
+        prev_block_finalization_proof: header_info.prev_block_finalization_proof,
+        previous_hash: last_header.hash(),
+        height: last_header.height + 1,
+        timestamp: header_info.timestamp,
+        tx_merkle_root: merkle_tree.root(),
+        state_merkle_root: storage.state_root().await?,
+        validator_set: delegation_state
+            .calculate_effective_validator_set()
+            .map_err(ExecutionError::InvalidState)?,
+        delegation_state_hash: delegation_state.hash(),
+        version: next_version,
+    })
+}
+
+async fn execute_add_validator(
+    storage: &mut StateStorage,
+    index: usize,
+    public_key: PublicKey,
+    voting_power: VotingPower,
+) -> Result<(), ExecutionError> {
+    let mut delegation_state = storage.get_delegation_state().await?;
+    if delegation_state
+        .original_validator_set
+        .iter()
+        .map(|(x, _, _)| x)
+        .any(|x| *x == public_key)
+    {
+        return Err(ExecutionError::TransactionFailure {
+            index,
+            message: "validator already exists".to_string(),
+        });
+    }
+    delegation_state
+        .original_validator_set
+        .push((public_key.clone(), voting_power, None));
+    storage.update_delegation_state(delegation_state).await?;
+    Ok(())
+}
+
+async fn execute_remove_validator(
+    storage: &mut StateStorage,
+    index: usize,
+    public_key: PublicKey,
+) -> Result<(), ExecutionError> {
+    let mut delegation_state = storage.get_delegation_state().await?;
+    let validator_index = delegation_state
+        .original_validator_set
+        .iter()
+        .position(|(x, _, _)| *x == public_key)
+        .ok_or_else(|| ExecutionError::TransactionFailure {
+            index,
+            message: "validator does not exist".to_string(),
+        })?;
+    delegation_state
+        .original_validator_set
+        .remove(validator_index);
```

Comment:
Because the delegatees are represented as `(first_delegatee: u64, current_delegatee: u64)`, these indices will point to the wrong nodes once we remove a node from `original_validator_set`.

Reply:
We should fix it here, or we should use `PublicKey` instead of `u64` when calculating the effective set of validators.

Reply:
@lemonholic It's handled [right below](https://github.com/postech-dao/simperby/pull/60/files#diff-bd1eb789c3438a54917a19f6c502e5d1af832b28c948f586c53171e2977d0099R172-R178).

Reply:
fixed.

Reply:
Suppose that a node has delegated its right to the node at the last index.
If we remove any node in the middle, the pointer(index) to the last node will be broken.

Reply:
fixed

Reply:
resolved by [0490d5e](https://github.com/postech-dao/simperby/pull/60/commits/0490d5e7b320e8c42d3b895dee74f62ffbe5ac10)